### PR TITLE
Overhaul the API to use Stores and StoreProviders.

### PR DESCRIPTION
**Pull request by chris-horner**
Modify converters to be more flexible.
Add converter for Moshi.

Fixes issue #18.

### CODE REVIEW

File: rxstore/src/main/java/au/com/gridstone/rxstore/StoreProvider.java [hunk 1]
```diff
@@ -0,0 +1,839 @@
...
+    if (!directory.exists()) {
+      if (!directory.mkdir()) throw new RuntimeException("Could not create directory for store.");
+    }
+
+    File file = new File(directory, key);
+
+    if (!file.exists()) {
+      try {
+        boolean created = file.createNewFile();
+        if (!created) throw new IOException("Could not create file for store.");
```
Comment (stphven) [inline rxstore/src/main/java/au/com/gridstone/rxstore/StoreProvider.java]: You use two lines here, but up at line 92 you do the same kind of thing in a single line.

Why not just `if (!file.createNewFile()) throw new IOException("Could not create file for store.");`?

File: rxstore/src/main/java/au/com/gridstone/rxstore/StoreProvider.java [hunk 2]
```diff
@@ -0,0 +1,839 @@
...
+     * a new instance through your StoreProvider.
+     * <p>
+     * The {@link Scheduler} used for this operation will be the one specified in
+     * {@link Builder#schedulingWith(Scheduler) schedulingWith()}.
+     */
+    @NotNull public Single<T> observeDelete() {
+      return Single.create(new Single.OnSubscribe<T>() {
+        @Override public void call(final SingleSubscriber<? super T> subscriber) {
+          try {
+            if (!file.exists()) throw new IOException("This store has been deleted.");
```
Comment (stphven) [inline rxstore/src/main/java/au/com/gridstone/rxstore/StoreProvider.java]: Might be a bit confusing if you call `.delete()` and you get an exception saying `This store has been deleted`. Maybe `This store has ALREADY been deleted` would be clearer.

File: rxstore/src/main/java/au/com/gridstone/rxstore/StoreProvider.java [hunk 3]
```diff
@@ -0,0 +1,839 @@
...
+            });
+          } catch (Exception e) {
+            subscriber.onError(e);
+          }
+        }
+      }).subscribeOn(scheduler);
+    }
+
+    /**
+     * Delete the file associated with this tore and render it completely unusable.
```
Comment (stphven) [inline rxstore/src/main/java/au/com/gridstone/rxstore/StoreProvider.java]: Missing an 's' from the start of 'store': `...associated with this TORE and render it...`

File: rxstore/src/main/java/au/com/gridstone/rxstore/StoreProvider.java [hunk 4]
```diff
@@ -0,0 +1,839 @@
...
+     * a new instance through your StoreProvider.
+     * <p>
+     * The {@link Scheduler} used for this operation will be the one specified in
+     * {@link Builder#schedulingWith(Scheduler) schedulingWith()}.
+     */
+    @NotNull public Single<List<T>> observeDelete() {
+      return Single.create(new Single.OnSubscribe<List<T>>() {
+        @Override public void call(final SingleSubscriber<? super List<T>> subscriber) {
+          try {
+            if (!file.exists()) throw new IOException("This store has been deleted.");
```
Comment (stphven) [inline rxstore/src/main/java/au/com/gridstone/rxstore/StoreProvider.java]: Again, calling `delete()` and getting an exception explaining that the store has been deleted is probably not the clearest.

File: rxstore/src/test/java/au/com/gridstone/rxstore/StoreProviderTest.java [hunk 5]
```diff
@@ -0,0 +1,406 @@
...
+
+  @Before public void setup() throws IOException {
+    storeProvider = StoreProvider.with(tempDir.newFolder("rxStoreTest"))
+        .schedulingWith(Schedulers.immediate())
+        .using(new TestConverter());
+  }
+
+  @Test public void putAndClear() {
+    ValueStore<TestData> valueStore = storeProvider.valueStore("testValue", TestData.class);
+    TestData value = new TestData("Test", 1);
```
Comment (stphven) [inline rxstore/src/test/java/au/com/gridstone/rxstore/StoreProviderTest.java]: In the previous tests you performed a check on the initial state to see if it was null, whereas in this test you don't. Just pointing out the inconsistency.

File: rxstore/src/test/java/au/com/gridstone/rxstore/StoreProviderTest.java [hunk 6]
```diff
@@ -0,0 +1,406 @@
...
+    TestData value2 = new TestData("Test2", 2);
+    valueStore.put(value2);
+    assertThat(observer.takeNext()).isEqualTo(value2);
+
+    valueStore.clear();
+    assertThat(observer.takeNext()).isNull();
+
+    valueStore.delete();
+    assertThat(observer.takeNext()).isNull();
+    observer.assertOnCompleted();
```
Comment (stphven) [inline rxstore/src/test/java/au/com/gridstone/rxstore/StoreProviderTest.java]: As far as I can tell, you're only asserting completeness here, at the very end. To prove that it's not complete BEFORE `.delete()` is called, consider putting some kind of `.assertNotCompleted()` call in before the `.delete()`.

File: rxstore/src/test/java/au/com/gridstone/rxstore/StoreProviderTest.java [hunk 7]
```diff
@@ -0,0 +1,406 @@
...
+  }
+
+  @Test public void putAndClearList() {
+    ListStore<TestData> store = storeProvider.listStore("testValues", TestData.class);
+    List<TestData> list = Arrays.asList(new TestData("Test1", 1), new TestData("Test2", 2));
```
Comment (stphven) [inline rxstore/src/test/java/au/com/gridstone/rxstore/StoreProviderTest.java]: Again, for consistency with the other tests, consider asserting that the store is empty before any data is inserted.

File: rxstore/src/test/java/au/com/gridstone/rxstore/StoreProviderTest.java [hunk 8]
```diff
@@ -0,0 +1,406 @@
...
+    List<TestData> expectedList = new ArrayList<>(list);
+    expectedList.add(newValue);
+    assertThat(observer.takeNext()).isEqualTo(expectedList);
+
+    store.clear();
+    assertThat(observer.takeNext()).isEmpty();
+
+    store.delete();
+    assertThat(observer.takeNext()).isEmpty();
+    observer.assertOnCompleted();
```
Comment (stphven) [inline rxstore/src/test/java/au/com/gridstone/rxstore/StoreProviderTest.java]: Again, consider asserting that the observer is NOT complete BEFORE `.delete()` is called, just to prove the change.

File: rxstore/src/test/java/au/com/gridstone/rxstore/StoreProviderTest.java (hunk 5 above)
Comment (chris-horner) [inline rxstore/src/test/java/au/com/gridstone/rxstore/StoreProviderTest.java]: I'm not sure I follow your meaning. I'm not check it's null initially here, and in the next test I'm *only* checking that the non-initialised store returns null.

File: rxstore/src/test/java/au/com/gridstone/rxstore/StoreProviderTest.java (hunk 7 above)
Comment (chris-horner) [inline rxstore/src/test/java/au/com/gridstone/rxstore/StoreProviderTest.java]: Which other tests are you referring to? There should only be two tests that do this: `getOnEmptyReturnsNull()` and `getOnEmptyListReturnsEmpty()`

File: README.md [hunk 9]
```diff
@@ -1,101 +1,97 @@
...
-	  .subscribeOn(Schedulers.io())
-	  .observeOn(AndroidSchedulers.mainThread())
-	  .subscribe((persistedDino) -> {
-	    // Do something with your persisted dino.
-	  });
+ValueStore<Person> store = storeProvider.valueStore("person", Person.class);
 ```
 
-If you're an [rxandroid][8] user, this code will look very familiar to you. The write operation will occur *off* the main thread, and we get informed of the completion back *on* the main thread.
+There are two ways we can add a `Person` to our store: `store.put(person)` or `store.observePut(person)`. `put()` is a fire-and-forget method that will asynchronously write the value to disk. `observePut()` returns an RxJava `Single` that must be subscribed to in order for the write operation begin. This is useful when incorporating the write operation into a chain, or would like to know when a write operation has completed.
```
Comment (jsprincep) [inline README.md]: missing a 'to' after operation.

File: rxstore/src/main/java/au/com/gridstone/rxstore/StoreProvider.java [hunk 10]
```diff
@@ -0,0 +1,838 @@
...
+ * </pre>
+ * Android users can also invoke...
+ * <pre>
+ * {@code
+ * StoreProvider storeProvider = StoreProvider.with(context).inDir("subDir").using(converter);
+ * }
+ * </pre>
+ * A StoreProvider can then be used to get a {@link ValueStore ValueStore} or a
+ * {@link ListStore ListStore} depending on your needs. ValueStores hold a single object,
+ * where as ListStores hold many (and have convenience methods for adding and removing single
```
Comment (jsprincep) [inline rxstore/src/main/java/au/com/gridstone/rxstore/StoreProvider.java]: whereas

File: rxstore/src/main/java/au/com/gridstone/rxstore/StoreProvider.java [hunk 11]
```diff
@@ -0,0 +1,838 @@
...
+ * storeProvider.listStore("someKey", Data.class);
+ * }
+ * </pre>
+ */
+public final class StoreProvider {
+  private final File directory;
+  private final Converter converter;
+  private final Scheduler scheduler;
+
+  private StoreProvider(File directory, Converter converter, Scheduler scheduler) {
```
Comment (jsprincep) [inline rxstore/src/main/java/au/com/gridstone/rxstore/StoreProvider.java]: Post-Kotlin,  this feels _so_ clunky.

Comment (chris-horner) [inline rxstore/src/main/java/au/com/gridstone/rxstore/StoreProvider.java]: In what way? Is there something about the builder API that you don't like? Happy to make changes now 😄

Comment (jsprincep) [inline rxstore/src/main/java/au/com/gridstone/rxstore/StoreProvider.java]: No. You can't, it's all Java's fault! ;) All I'm talking about is the the whole this.thing = thing in the constructor.

**Comment by jsprincep** [comment]
Apart from my grammar-nazi pickups - looks good! Nice work!

**Comment by stphven** [comment]
LG2M

**Pull request merged**